objectherkenning_openbare_ruimte/on_edge/data_delivery_pipeline/components/data_delivery.py

```diff
@@ -0,0 +1,316 @@
+    get_img_name_from_csv_row,
+    log_execution_time,
+    save_csv_file,
+)
+from objectherkenning_openbare_ruimte.settings.settings import (
+    ObjectherkenningOpenbareRuimteSettings,
+)
+
+logger = logging.getLogger("data_delivery_pipeline")
+
+
+class DataDelivery:
+    def __init__(self):
+        self.settings = ObjectherkenningOpenbareRuimteSettings.get_settings()
+        self.detections_folder = self.settings["data_delivery_pipeline"][
+            "detections_path"
+        ]
+        self.metadata_folder = self.settings["data_delivery_pipeline"]["metadata_path"]
+        self.metadata_csv_file_paths_with_errors: List[str] = []
+        self.model_and_code_version = [
+            self.settings["detection_pipeline"]["model_name"],
+            self.settings["data_delivery_pipeline"]["ml_model_id"],
+            self.settings["data_delivery_pipeline"]["project_version"],
+        ]
+        self.detection_metadata_header = [
+            "image_name",
+            "object_class",
+            "x_center",
+            "y_center",
+            "width",
+            "height",
+            "confidence",
+            "tracking_id",
+        ]
+
+    @log_execution_time
+    def run_pipeline(self):
+        """
+        Runs the data delivery pipeline:
+            - retrieves all the csvs and images that need to be delivered;
+            - delivers the data to Azure;
+            - deletes the delivered data.
+        """
+        logger.debug(f"Running delivery pipeline on {self.detections_folder}..")
+        metadata_csv_file_paths = get_frame_metadata_csv_file_paths(
+            root_folder=self.detections_folder
+        )
+        logger.info(f"Number of CSVs to deliver: {len(metadata_csv_file_paths)}")
+
+        for frame_metadata_file_path in metadata_csv_file_paths:
+            self._deliver_data(frame_metadata_file_path=frame_metadata_file_path)
+            self._delete_processed_data(metadata_csv_file_path=frame_metadata_file_path)
+
+    @log_execution_time
+    def _deliver_data(self, frame_metadata_file_path):
+        """
+        Using Azure IoT delivers the images and metadata to Azure.
+
+        Parameters
+        ----------
+        frame_metadata_file_paths
+            CSV files containing the metadata of the pictures,
+            it's used to keep track of which files need to be delivered.
+
+        """
+        iot_handler = IoTHandler(
+            hostname=self.settings["azure_iot"]["hostname"],
+            device_id=self.settings["azure_iot"]["device_id"],
+            shared_access_key=self.settings["azure_iot"]["shared_access_key"],
+        )
+        try:
+            self._deliver_data_batch(frame_metadata_file_path, iot_handler)
+            if frame_metadata_file_path in self.metadata_csv_file_paths_with_errors:
+                self.metadata_csv_file_paths_with_errors.remove(
+                    frame_metadata_file_path
+                )
+        except FileNotFoundError as e:
```

Comment:
Yes, it's a strange situation. It means that the previous function found the file and added it to a list, but someone in the meantime deleted it. Do you think it can happen?
I cannot move the metadata and the images because I don't know them, since I cannot open the file and check. And the file itself doesn't exists so of course I cannot move neither that.

Reply:
Ah yes if the whole metadata file is not there, then there's nothing we can do. I was more thinking if a single image or annotation.txt is missing, then also the error is raised and the processing of that metadata file stops.

Reply:
But... I also don't see how this could happen other than someone manually deleting a file. So let's not overthink it :)